### PR TITLE
Add PodExecOptions #2332

### LINKED ISSUE #2332: PodExecOptions (as object of AdmissionReview#request) not supported?
When reading JSON representing an AdmissionReview for a pod/exec request (**PodExecOptions**), no resource type is found.

With this code (which can be cut down to a simple test if needed):
```
...
// imports included for completeness
import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.fabric8.kubernetes.api.model.admission.AdmissionRequest;
import io.fabric8.kubernetes.api.model.admission.AdmissionResponseBuilder;
import io.fabric8.kubernetes.api.model.admission.AdmissionReview;
...
	@POST
	public AdmissionReview validate(String admissionReviewJSON) {
		ObjectMapper objectMapper = new ObjectMapper();
		final AdmissionReview admissionReview;
		try {
			// have to use this approach as passing parameter of type AdmissionReview doesn't work
			admissionReview = objectMapper.readValue(admissionReviewJSON, AdmissionReview.class);
		} catch (JsonProcessingException e) {
			logger.error("Failed to parse AdmissionReview", e);
			return null;
		}

		admissionReview.setResponse(new AdmissionResponseBuilder()
				.withUid(admissionRequest.getUid())
				.withAllowed(true) // TODO: check
				.build());
		return admissionReview;
	}
```

The JSON passed in is:
```
{
   "kind":"AdmissionReview",
   "apiVersion":"admission.k8s.io/v1",
   "request":{
      "uid":"b7a46c3d-a6c8-4f57-b616-27d839f4e210",
      "kind":{
         "group":"",
         "version":"v1",
         "kind":"PodExecOptions"
      },
      "resource":{
         "group":"",
         "version":"v1",
         "resource":"pods"
      },
      "subResource":"exec",
      "requestKind":{
         "group":"",
         "version":"v1",
         "kind":"PodExecOptions"
      },
      "requestResource":{
         "group":"",
         "version":"v1",
         "resource":"pods"
      },
      "requestSubResource":"exec",
      "name":"dummy-599f7fdbdc-22gmf",
      "namespace":"test123",
      "operation":"CONNECT",
      "userInfo":{
         "username":"admin",
         "uid":"admin",
         "groups":[
            "system:masters",
            "system:authenticated"
         ]
      },
      "object":{
         "kind":"PodExecOptions",
         "apiVersion":"v1",
         "stdin":true,
         "stdout":true,
         "tty":true,
         "container":"dummy",
         "command":[
            "sh",
            "-c",
            "command -v bash \u003e/dev/null \u0026\u0026 exec bash || exec sh"
         ]
      },
      "oldObject":null,
      "dryRun":false,
      "options":null
   }
}
```

I get the following error:
`Failed to parse AdmissionReview: com.fasterxml.jackson.databind.JsonMappingException: No resource type found for:v1#PodExecOptions`

What is the expected approach here?
Is it simply that the model is missing and might ship with a new release in future?
Or am I missing something? (If so, what?)

### PR DESCRIPTION
**Pull request by duco-lw**
## Description

Fix #2332 
Adds PodExecOptions model.
(Also fixes small typos in a couple of READMEs).
Let me know if I've missed something...

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by rohanKanojia** [comment]
ok to test

File: kubernetes-model-generator/README.md [hunk 1]
```diff
@@ -62,7 +62,7 @@ If everything works well, you would have model upgraded to specified Kubernetes/
  - `cmd/generate/generate.go` contains the main entry point and the descriptor of the schema.
    Add new entries to the `Schema` struct and `schemagen.PackageDescriptor` to generate
    an entry for any additional Go structs.
- - `pkg/schenage` package contains the source files with the logic to generate the schema.
+ - `pkg/schemage` package contains the source files with the logic to generate the schema.
```
Comment (rohanKanojia) [inline kubernetes-model-generator/README.md:65]: Shouldn't this be `pkg/schemagen`?

https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-model-generator/pkg/schemagen

**Comment by rohanKanojia** [comment]
@duco-lw : Thanks for your contribution :heart: . Were you able to resolve deserialization error regarding AdmissionReview using SNAPSHOT version of this fix? Maybe we could add a test for that here? https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/AdmissionReviewTest.java

**Comment by duco-lw** [comment]
@rohanKanojia I've made a test, however... this opened a can of worms...

`AdmissionRequest.object` is a `HasMetadata`, but `PodExecOptions` is a `KubernetesResource` (causing unmarshelling error) : https://github.com/fabric8io/kubernetes-client/blob/7868b41d47f099493c62c47279db8c67cc0bc9e3/kubernetes-model-generator/kubernetes-model-admissionregistration/src/main/java/io/fabric8/kubernetes/api/model/admission/AdmissionRequest.java#L74-L78

I manually changed this to a `KubernetesResource`, and the JSON could be parsed... :+1: 
But I figure this isn't the correct approach to fixing this... :-1: 

Looking at: https://github.com/fabric8io/kubernetes-client/blob/7868b41d47f099493c62c47279db8c67cc0bc9e3/kubernetes-model-generator/pkg/schemagen/generate.go#L166-L176

https://godoc.org/k8s.io/api/admission/v1#AdmissionRequest shows us that `object` is a `RawExtension`...
And https://godoc.org/k8s.io/api/core/v1#PodExecOptions gets generates as a `KubernetesResource` (not a `HasMetadata`) - because, while it does have `kind` and `apiVersion`, it doesn't have labels, etc...

I imagine the below change is out of scope of this PR, and likely in need of further discussion (I can see comments in code mentioning RawExtension):
```
		case "RawExtension":
-			return "io.fabric8.kubernetes.api.model.HasMetadata"
+			return "io.fabric8.kubernetes.api.model.KubernetesResource"
```

Perhaps some middle type between `KubernetesResource` and `HasMetadata`?
Something like:
```
public interface VersionedResource extends KubernetesResource {

  String getKind();

  String getApiVersion();
  void setApiVersion(String version);
}

-public interface HasMetadata extends KubernetesResource {
+public interface HasMetadata extends VersionedResource {

  ObjectMeta getMetadata();
  void setMetadata(ObjectMeta metadata);

-  String getKind();

-  String getApiVersion();
-  void setApiVersion(String version);
}
```

TL;DR: this PR is probably fine as it is, but doesn't get `PodExecOptions` working with `AdmissionRequest` because of `RawExtension` being modelled as `HasMetadata` (where `PodExecOptions` is a `KubernetesResource`)

**Comment by rohanKanojia** [comment]
Hmm, I remember I did something similar in https://github.com/fabric8io/kubernetes-client/pull/2073 but with options field(this is the reason we have our own `AdmissionRequest` class instead of a generated one). Do you know what possible values can object have?

**Comment by duco-lw** [comment]
I cannot list all possible types that `AdmissionRequest.object` and `AdmissionRequest.oldObject` could be - only that they are of type `RawExtension`...
I'd suggest changing them from `HasMetadata` to `KubernetesResource` as this seems to fix the problem for me... but I'm not confident it won't have impact elsewhere.
I'll add this to this PR - I did not realise `AdmissionRequest` was not generated...

**Comment by rohanKanojia** [comment]
[merge]

**Comment by rohanKanojia** [comment]
[merge]

**Comment by rohanKanojia** [comment]
retest this please

**Comment by rohanKanojia** [comment]
[merge]

**Comment by rohanKanojia** [comment]
There seems to be some problem with our Jenkins merge job :sob:

**Comment by rohanKanojia** [comment]
[merge]

**Comment by rohanKanojia** [comment]
[merge]

**Comment by duco-lw** [comment]
Pending reviewer perhaps?

**Comment by rohanKanojia** [comment]
let me merge using my admin privileges
